File: backend/app/api/scans.py
"""
Scans API endpoints
"""
from fastapi import APIRouter, Depends, HTTPException, BackgroundTasks
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, desc
from typing import List, Optional
from pydantic import BaseModel, Field
from uuid import UUID
from datetime import datetime

from app.core.database import get_db
from app.models.scan import Scan, ScanResult, ScanLog
from app.services.scanner import scanner

router = APIRouter()

# Pydantic schemas
class ScanConfiguration(BaseModel):
    """Scan configuration schema"""
    timeout: Optional[int] = Field(default=1800, description="Scan timeout in seconds")
    max_hosts: Optional[int] = Field(default=256, description="Maximum number of hosts")
    timing: Optional[str] = Field(default="T4", description="Nmap timing template")
    additional_args: Optional[str] = Field(default="", description="Additional nmap arguments")

class ScanCreate(BaseModel):
    """Create scan request schema"""
    name: str = Field(..., description="Scan name")
    target: str = Field(..., description="Target IP, hostname, or network (e.g., 192.168.1.0/24)")
    scan_type: str = Field(..., description="Scan type (quick, full, service, etc.)")
    nmap_arguments: Optional[str] = Field(default=None, description="Custom nmap arguments")
    configuration: Optional[ScanConfiguration] = None

class ScanResponse(BaseModel):
    """Scan response schema"""
    id: UUID
    name: str
    target: str
    scan_type: str
    status: str
    progress: int
    created_at: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    error_message: Optional[str] = None

    class Config:
        from_attributes = True

class ScanResultResponse(BaseModel):
    """Scan result response schema"""
    id: UUID
    scan_id: UUID
    host: str
    hostname: Optional[str]
    state: Optional[str]
    ports: Optional[list]
    os_detection: Optional[dict]
    services: Optional[list]
    mac_address: Optional[str]
    mac_vendor: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True

class ScanLogResponse(BaseModel):
    """Scan log response schema"""
    id: UUID
    level: str
    message: str
    created_at: datetime

    class Config:
        from_attributes = True

async def run_scan_background(scan_id: str, target: str, arguments: str, db: AsyncSession):
    """Background task to run scan"""
    try:
        await scanner.execute_scan(scan_id, target, arguments, db)
    except Exception as e:
        print(f"Scan {scan_id} failed: {str(e)}")

@router.post("/", response_model=ScanResponse)
async def create_scan(
    scan_data: ScanCreate,
    background_tasks: BackgroundTasks,
    db: AsyncSession = Depends(get_db)
):
    """
    Create and start a new scan

    Args:
        scan_data: Scan configuration
        background_tasks: FastAPI background tasks
        db: Database session

    Returns:
        Created scan object
    """
    # Determine nmap arguments
    if scan_data.nmap_arguments:
        nmap_args = scan_data.nmap_arguments
    else:
        templates = scanner.get_scan_templates()
        if scan_data.scan_type in templates:
            nmap_args = templates[scan_data.scan_type]["arguments"]
        else:
            raise HTTPException(status_code=400, detail=f"Unknown scan type: {scan_data.scan_type}")

    # Create scan record
    scan = Scan(
        name=scan_data.name,
        target=scan_data.target,
        scan_type=scan_data.scan_type,
        status="pending",
        configuration=scan_data.configuration.model_dump() if scan_data.configuration else None
    )

    db.add(scan)
    await db.commit()
    await db.refresh(scan)

    # Start scan in background
    background_tasks.add_task(
        run_scan_background,
        str(scan.id),
        scan_data.target,
        nmap_args,
        db
    )

    return scan

@router.get("/", response_model=List[ScanResponse])
async def list_scans(
    skip: int = 0,
    limit: int = 100,
    status: Optional[str] = None,
    db: AsyncSession = Depends(get_db)
):
    """
    List all scans with optional filtering

    Args:
        skip: Number of records to skip
        limit: Maximum number of records to return
        status: Filter by status (pending, running, completed, failed)
        db: Database session

    Returns:
        List of scans
    """
    query = select(Scan).order_by(desc(Scan.created_at))

    if status:
        query = query.where(Scan.status == status)

    query = query.offset(skip).limit(limit)

    result = await db.execute(query)
    scans = result.scalars().all()

    return scans

@router.get("/{scan_id}", response_model=ScanResponse)
async def get_scan(
    scan_id: UUID,
    db: AsyncSession = Depends(get_db)
):
    """
    Get scan details by ID

    Args:
        scan_id: Scan UUID
        db: Database session

    Returns:
        Scan object
    """
    result = await db.execute(select(Scan).where(Scan.id == scan_id))
    scan = result.scalar_one_or_none()

    if not scan:
        raise HTTPException(status_code=404, detail="Scan not found")

    return scan

@router.get("/{scan_id}/results", response_model=List[ScanResultResponse])
async def get_scan_results(
    scan_id: UUID,
    db: AsyncSession = Depends(get_db)
):
    """
    Get scan results by scan ID

    Args:
        scan_id: Scan UUID
        db: Database session

    Returns:
        List of scan results
    """
    result = await db.execute(
        select(ScanResult).where(ScanResult.scan_id == scan_id)
    )
    results = result.scalars().all()

    return results

@router.get("/{scan_id}/logs", response_model=List[ScanLogResponse])
async def get_scan_logs(
    scan_id: UUID,
    db: AsyncSession = Depends(get_db)
):
    """
    Get scan logs by scan ID

    Args:
        scan_id: Scan UUID
        db: Database session

    Returns:
        List of scan logs
    """
    result = await db.execute(
        select(ScanLog)
        .where(ScanLog.scan_id == scan_id)
        .order_by(ScanLog.created_at)
    )
    logs = result.scalars().all()

    return logs

@router.delete("/{scan_id}")
async def delete_scan(
    scan_id: UUID,
    db: AsyncSession = Depends(get_db)
):
    """
    Delete a scan and all its results

    Args:
        scan_id: Scan UUID
        db: Database session

    Returns:
        Success message
    """
    result = await db.execute(select(Scan).where(Scan.id == scan_id))
    scan = result.scalar_one_or_none()

    if not scan:
        raise HTTPException(status_code=404, detail="Scan not found")

    await db.delete(scan)
    await db.commit()

    return {"message": "Scan deleted successfully"}

@router.post("/{scan_id}/cancel")
async def cancel_scan(
    scan_id: UUID,
    db: AsyncSession = Depends(get_db)
):
    """
    Cancel a running scan

    Args:
        scan_id: Scan UUID
        db: Database session

    Returns:
        Success message
    """
    result = await db.execute(select(Scan).where(Scan.id == scan_id))
    scan = result.scalar_one_or_none()

    if not scan:
        raise HTTPException(status_code=404, detail="Scan not found")

    if scan.status != "running":
        raise HTTPException(status_code=400, detail="Scan is not running")

    scanner.cancel_scan(str(scan_id))

    scan.status = "cancelled"
    await db.commit()

    return {"message": "Scan cancelled"}

File: backend-go/internal/api/handlers/scans.go
package handlers

import (
	"context"
	"fmt"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
	"github.com/nmap-scanner/backend-go/internal/database"
	"github.com/nmap-scanner/backend-go/internal/models"
	"github.com/nmap-scanner/backend-go/internal/scanner"
)

type ScanHandler struct {
	db      *database.Database
	scanner *scanner.Scanner
}

func NewScanHandler(db *database.Database, scanner *scanner.Scanner) *ScanHandler {
	return &ScanHandler{
		db:      db,
		scanner: scanner,
	}
}

// CreateScan creates and starts a new scan
func (h *ScanHandler) CreateScan(c *fiber.Ctx) error {
	var req models.CreateScanRequest
	if err := c.BodyParser(&req); err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "Invalid request body"})
	}

	// Validate required fields
	if req.Name == "" || req.Target == "" || req.ScanType == "" {
		return c.Status(400).JSON(fiber.Map{"error": "name, target, and scan_type are required"})
	}

	// Determine nmap arguments
	nmapArgs := ""
	if req.NmapArguments != nil {
		nmapArgs = *req.NmapArguments
	} else {
		templates := h.scanner.GetScanTemplates()
		if template, ok := templates[req.ScanType]; ok {
			nmapArgs = template["arguments"]
		} else {
			return c.Status(400).JSON(fiber.Map{"error": fmt.Sprintf("Unknown scan type: %s", req.ScanType)})
		}
	}

	// Create scan record
	scanID := uuid.New()
	query := `
		INSERT INTO scans (id, name, target, scan_type, status, progress, created_at, configuration)
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
		RETURNING id, name, target, scan_type, status, progress, created_at
	`

	var scan models.Scan
	err := h.db.Pool.QueryRow(context.Background(), query,
		scanID, req.Name, req.Target, req.ScanType, "pending", 0, time.Now(), req.Configuration,
	).Scan(&scan.ID, &scan.Name, &scan.Target, &scan.ScanType, &scan.Status, &scan.Progress, &scan.CreatedAt)

	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to create scan"})
	}

	// Start scan in background
	go func() {
		ctx := context.Background()
		if err := h.scanner.ExecuteScan(ctx, scanID, req.Target, nmapArgs); err != nil {
			fmt.Printf("Scan %s failed: %v\n", scanID, err)
		}
	}()

	return c.Status(201).JSON(scan)
}

// ListScans returns all scans
func (h *ScanHandler) ListScans(c *fiber.Ctx) error {
	query := `
		SELECT id, name, target, scan_type, status, progress, created_at, started_at, completed_at, error_message
		FROM scans
		ORDER BY created_at DESC
		LIMIT 100
	`

	rows, err := h.db.Pool.Query(context.Background(), query)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to fetch scans"})
	}
	defer rows.Close()

	scans := []models.Scan{}
	for rows.Next() {
		var scan models.Scan
		err := rows.Scan(&scan.ID, &scan.Name, &scan.Target, &scan.ScanType, &scan.Status,
			&scan.Progress, &scan.CreatedAt, &scan.StartedAt, &scan.CompletedAt, &scan.ErrorMessage)
		if err != nil {
			continue
		}
		scans = append(scans, scan)
	}

	return c.JSON(scans)
}

// GetScan returns a specific scan by ID
func (h *ScanHandler) GetScan(c *fiber.Ctx) error {
	scanID := c.Params("id")

	query := `
		SELECT id, name, target, scan_type, status, progress, created_at, started_at, completed_at, error_message
		FROM scans
		WHERE id = $1
	`

	var scan models.Scan
	err := h.db.Pool.QueryRow(context.Background(), query, scanID).Scan(
		&scan.ID, &scan.Name, &scan.Target, &scan.ScanType, &scan.Status,
		&scan.Progress, &scan.CreatedAt, &scan.StartedAt, &scan.CompletedAt, &scan.ErrorMessage,
	)

	if err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "Scan not found"})
	}

	return c.JSON(scan)
}

// GetScanResults returns results for a specific scan
func (h *ScanHandler) GetScanResults(c *fiber.Ctx) error {
	scanID := c.Params("id")

	query := `
		SELECT id, scan_id, host, hostname, state, ports, os_detection, services, mac_address, mac_vendor, created_at
		FROM scan_results
		WHERE scan_id = $1
	`

	rows, err := h.db.Pool.Query(context.Background(), query, scanID)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to fetch results"})
	}
	defer rows.Close()

	results := []models.ScanResult{}
	for rows.Next() {
		var result models.ScanResult
		err := rows.Scan(&result.ID, &result.ScanID, &result.Host, &result.Hostname, &result.State,
			&result.Ports, &result.OSDetection, &result.Services, &result.MacAddress, &result.MacVendor, &result.CreatedAt)
		if err != nil {
			continue
		}
		results = append(results, result)
	}

	return c.JSON(results)
}

// GetScanLogs returns logs for a specific scan
func (h *ScanHandler) GetScanLogs(c *fiber.Ctx) error {
	scanID := c.Params("id")

	query := `
		SELECT id, scan_id, level, message, created_at
		FROM scan_logs
		WHERE scan_id = $1
		ORDER BY created_at ASC
	`

	rows, err := h.db.Pool.Query(context.Background(), query, scanID)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to fetch logs"})
	}
	defer rows.Close()

	logs := []models.ScanLog{}
	for rows.Next() {
		var log models.ScanLog
		err := rows.Scan(&log.ID, &log.ScanID, &log.Level, &log.Message, &log.CreatedAt)
		if err != nil {
			continue
		}
		logs = append(logs, log)
	}

	return c.JSON(logs)
}
